### ISSUE TITLE
OrientDB isn't doing short-circuit evaluation

### ISSUE THREAD

**Issue by anjosc** [orientechnologies/orientdb#3904]
I'm getting much worse results with simple query like this:
```sql
select from table1
where link.value = 'somevalue'
and 'othervalue' in linkset.deep.expensive.nested.relationship.value
```
Than with this:
```sql
select from (
  select from table1
  where link.value = 'somevalue'
) where 'othervalue' in linkset.deep.expensive.nested.relationship.value
```
I should get similar performances if, as soon as the first condition fails (is false), the second one does not need to be evaluated (and traversed)

Also, when using a condition with "IN" as in the example, we could stop traversing as soon as the condition is met, but given the performance I'm getting, I think it's evaluating everything (and traversing) and then checking if the condition matches.
BTW, I get the same performance wether I use "'othervalue' in linkset.deep.expensive.nested.relationship.value" or "linkset.deep.expensive.nested.relationship.value contains 'othervalue'"

**Comment by lvca** [comment]
Are you running these micro-benchmarks form console or studio? if yes, assure you have the same limits. Try these:
```sql
select from table1
where link.value = 'somevalue'
and 'othervalue' in linkset.deep.expensive.nested.relationship.value
limit -1
```

Than with this:

```sql
select from (
  select from table1
  where link.value = 'somevalue' limit -1
) where 'othervalue' in linkset.deep.expensive.nested.relationship.value limit -1
```

**Comment by anjosc** [comment]
Hi @lvca 
I just tried that with similar results.
Also if I change the order of conditions in the 'where' clause in the first example, I get the exact same performance.

**Comment by lvca** [comment]
I see you are using Document API. If you switch to GraphAPI the fastest way to do this is lookup one (or more) vertices and the start traversing. Example:

```sql
select in('deep').in('expensive').in('nested').in('relationship')
from table33
where  'othervalue' in value
```

In this way you lookup the value (end of the cain of your example) and you go back.

**Comment by anjosc** [comment]
Thanks for looking into this.
However the purpose was to be able to get fast responses without having to fine-tune individual queries.
That's the role of a query planner/optimizer in RDMS.
What I suspect is that, rather than not doing short-circuit-evalution as I initially thought, OrientDB lacks an advanced Query Planner.
For our purposes, which is a typical RDBMS scenario with some JSON in the mix, even though doing joins is O(log n) and not O(1) as you imply, we got better performance in every benchmark we made, using PostgreSQL.
So, I'm afraid we will no be using OrientDB at this moment.
I'll keep an aye in the project though.
Best regards
